codeguru_profiler_agent/profiler_disabler.py

```diff
@@ -38,19 +42,39 @@ class CpuUsageCheck:
     def __init__(self, timer):
         self.timer = timer
 
-    def is_cpu_usage_limit_reached(self, profile=None):
+    # This function carries out an overall cpu limit check that covers the cpu overhead caused for the full
+    # sampling cycle: sample -> aggregate -> report -> refresh config. We expect this function to be called after
+    # configuration refresh and profile submission.
+    def is_overall_cpu_usage_limit_reached(self, profile=None):
         profiler_metric = self.timer.metrics.get("runProfiler")
-        if not profiler_metric or profiler_metric.counter < MINIMUM_MEASURES_IN_DURATION_METRICS:
+        if not profile or not profiler_metric or profiler_metric.counter < MINIMUM_MEASURES_IN_DURATION_METRICS:
+            return False
+
+        used_time_percentage = 100 * profiler_metric.total/(profile.get_active_millis_since_start()/1000)
+
+        if used_time_percentage >= AgentConfiguration.get().cpu_limit_percentage:
+            logger.debug(self.timer.metrics)
+            logger.info(
+                "Profiler overall cpu usage limit reached: {:.2f} % (limit: {:.2f} %), will stop CodeGuru Profiler."
+                .format(used_time_percentage, AgentConfiguration.get().cpu_limit_percentage))
```

Comment:
it is probably not possible with current code but if `AgentConfiguration.get().cpu_limit_percentage` is changed between the test and the log it would make a hard bug to understand. maybe use a local variable, same for next function. I realize this was like that before you touched it.

Reply:
Good point!